File: may/05/centa.py
# O(n) runtime and O(n) space complexity
# uniquely identify each node with and integer and add
# the ones that appear to the code
# used recursion but an iterative solution is possible
# by: Centa

from collections import deque


class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def _serialize(root, id, code_builder):
    if root:
        code_builder.append("{:}:{:}".format(id, root.val))
        _serialize(root.left, 2*id, code_builder)
        _serialize(root.right, 2*id+1, code_builder)


def serialize(root):
    code_builder = []

    _serialize(root, 1, code_builder)

    return " ".join(code_builder)


def _deserialize(id, node_map):
    if id in node_map.keys():
        node = Node(node_map[id])
        node.left = _deserialize(2*id, node_map)
        node.right = _deserialize(2*id+1, node_map)
        return node
    return None


def deserialize(code):
    node_map = { int(s.split(":")[0]):s.split(":")[1] for s in code.split(" ") }

    return _deserialize(1, node_map)


if __name__ == "__main__":
    # driver code
    node = Node('root', Node('left', Node('left.left')), Node('right'))

    assert deserialize(serialize(node)).left.left.val == 'left.left'

File: may/10/centa.cc
// O(n) runtime and O(1) space complexity with post order traversal
#include <iostream>

struct Node {
    int data;
    Node *left, *right;

    Node(int data) : data(data) {
        left = right = NULL;
    }

    Node(int data, Node *left, Node *right) :
        data(data), left(left), right(right) { }
};

bool check(Node *root, int& count) {
    bool ans = true;

    if (root->left != NULL &&
        (!check(root->left, count) ||
        root->left->data != root->data))
        ans = false;

    if (root->right != NULL &&
        (!check(root->right, count) ||
        root->right->data != root->data))
        ans = false;

    if (ans) ++count;

    return ans;
}

int count_universal_subtrees(Node *root) {
    int count = 0;

    if (root != NULL) check(root, count);

    return count;
}

int main() {
    Node *root  = new Node(0, new Node(1), new Node(0, new Node(1 , new Node(1), new Node(1)), new Node(0)));

    std::cout << count_universal_subtrees(root) << std::endl;

    return 0;
}